### ISSUE TITLE
Image mod operation results in a longer response time due to multiple retries

### ISSUE THREAD

**Issue by airycanon** [regclient/regclient#384]
### Description

Performing an image mod operation with v0.4.7 results in a longer response time compared to v0.4.5 due to multiple retries.

### Current Behavior

outputs with regctl v0.4.7, the real domain name has been hidden.
```
>>> regctl image mod --replace --annotation a=b <registry>/e2e-automation/deploy:deploy --verbosity debug
DEBU[0000] Loading docker config                         api= blobChunk=0 blobMax=0 helper=docker-credential-desktop hostname=<registry> mirrors="[]" name=<registry> pathPrefix= repoAuth=false tls=enabled user=
DEBU[0000] Loading docker config                         api= blobChunk=0 blobMax=0 helper=docker-credential-desktop hostname=registry-1.docker.io mirrors="[]" name=docker.io pathPrefix= repoAuth=false tls=enabled user=
DEBU[0000] Loading host config                           api= blobChunk=0 blobMax=0 helper= hostname=<registry> mirrors="[]" name=<registry> pathPrefix= repoAuth=false tls=disabled user=
WARN[0000] Changing TLS settings for registry            host=<registry> new=disabled orig=enabled
DEBU[0000] regclient initialized                         VCSRef=f43ada160d953317c51e7d9170afa2a316f4b9be VCSTag=
DEBU[0000] Modifying image                               ref="<registry>/e2e-automation/deploy:deploy"
DEBU[0000] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/manifests/deploy" withAuth=false
DEBU[0000] Auth request parsed                           challenge="[{bearer map[realm:http://<registry>/service/token scope:repository:e2e-automation/deploy:pull service:harbor-registry]}]"
DEBU[0000] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/manifests/deploy" withAuth=true
DEBU[0000] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/manifests/sha256:ab4c70e7d724f6de17e996fa2e8c0e01ae65802eae3b29b29591313c1f1063ce" withAuth=true
DEBU[0001] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/blobs/sha256:1b60b809c9a33d4410ffef20b04961eadd3cd4a669bd5cfe2a79ce8f76408031" withAuth=true
DEBU[0001] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/referrers/sha256:ab4c70e7d724f6de17e996fa2e8c0e01ae65802eae3b29b29591313c1f1063ce" withAuth=true
DEBU[0001] Auth request parsed                           challenge="[]"
WARN[0001] Failed to handle auth request                 Err="empty challenge header" URL="http://<registry>/v2/e2e-automation/deploy/referrers/sha256:ab4c70e7d724f6de17e996fa2e8c0e01ae65802eae3b29b29591313c1f1063ce"
WARN[0001] Sleeping for backoff                          Host=<registry> Seconds=1.999932792
DEBU[0003] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/_oci/artifacts/referrers?digest=sha256%3Aab4c70e7d724f6de17e996fa2e8c0e01ae65802eae3b29b29591313c1f1063ce" withAuth=true
DEBU[0003] Auth request parsed                           challenge="[]"
WARN[0003] Failed to handle auth request                 Err="empty challenge header" URL="http://<registry>/v2/e2e-automation/deploy/_oci/artifacts/referrers?digest=sha256%3Aab4c70e7d724f6de17e996fa2e8c0e01ae65802eae3b29b29591313c1f1063ce"
WARN[0003] Sleeping for backoff                          Host=<registry> Seconds=3.9998725410000002
DEBU[0007] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/manifests/sha256-ab4c70e7d724f6de17e996fa2e8c0e01ae65802eae3b29b29591313c1f1063ce" withAuth=true
DEBU[0007] Request failed                                Status="Not Found" URL="http://<registry>/v2/e2e-automation/deploy/manifests/sha256-ab4c70e7d724f6de17e996fa2e8c0e01ae65802eae3b29b29591313c1f1063ce"
DEBU[0007] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/manifests/sha256:4a8227617ca11e623284852400c12cf2f30bf991597deac1abbc9f7d3927e22b" withAuth=true
DEBU[0008] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/blobs/sha256:ee9f4d7c388eafb922afbca879b1e3874e091a7b5a4e4be0546c93ce3a12b7cb" withAuth=true
DEBU[0009] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/referrers/sha256:4a8227617ca11e623284852400c12cf2f30bf991597deac1abbc9f7d3927e22b" withAuth=true
DEBU[0009] Auth request parsed                           challenge="[]"
WARN[0009] Failed to handle auth request                 Err="empty challenge header" URL="http://<registry>/v2/e2e-automation/deploy/referrers/sha256:4a8227617ca11e623284852400c12cf2f30bf991597deac1abbc9f7d3927e22b"
WARN[0009] Sleeping for backoff                          Host=<registry> Seconds=1.999892459
DEBU[0011] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/_oci/artifacts/referrers?digest=sha256%3A4a8227617ca11e623284852400c12cf2f30bf991597deac1abbc9f7d3927e22b" withAuth=true
DEBU[0011] Auth request parsed                           challenge="[]"
WARN[0011] Failed to handle auth request                 Err="empty challenge header" URL="http://<registry>/v2/e2e-automation/deploy/_oci/artifacts/referrers?digest=sha256%3A4a8227617ca11e623284852400c12cf2f30bf991597deac1abbc9f7d3927e22b"
WARN[0011] Sleeping for backoff                          Host=<registry> Seconds=3.999947292
DEBU[0015] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/manifests/sha256-4a8227617ca11e623284852400c12cf2f30bf991597deac1abbc9f7d3927e22b" withAuth=true
DEBU[0015] Request failed                                Status="Not Found" URL="http://<registry>/v2/e2e-automation/deploy/manifests/sha256-4a8227617ca11e623284852400c12cf2f30bf991597deac1abbc9f7d3927e22b"
DEBU[0015] http req                                      method=HEAD url="http://<registry>/v2/e2e-automation/deploy/manifests/deploy" withAuth=true
DEBU[0015] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/referrers/sha256:3e4b79029ce4089f26d78c8f3d98d936f1de7eb4051a251457b2a1ec01b849e2" withAuth=true
DEBU[0015] Auth request parsed                           challenge="[]"
WARN[0015] Failed to handle auth request                 Err="empty challenge header" URL="http://<registry>/v2/e2e-automation/deploy/referrers/sha256:3e4b79029ce4089f26d78c8f3d98d936f1de7eb4051a251457b2a1ec01b849e2"
WARN[0015] Sleeping for backoff                          Host=<registry> Seconds=3.999985333
DEBU[0019] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/_oci/artifacts/referrers?digest=sha256%3A3e4b79029ce4089f26d78c8f3d98d936f1de7eb4051a251457b2a1ec01b849e2" withAuth=true
DEBU[0020] Auth request parsed                           challenge="[]"
WARN[0020] Failed to handle auth request                 Err="empty challenge header" URL="http://<registry>/v2/e2e-automation/deploy/_oci/artifacts/referrers?digest=sha256%3A3e4b79029ce4089f26d78c8f3d98d936f1de7eb4051a251457b2a1ec01b849e2"
WARN[0020] Sleeping for backoff                          Host=<registry> Seconds=7.999943041
DEBU[0028] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/manifests/sha256-3e4b79029ce4089f26d78c8f3d98d936f1de7eb4051a251457b2a1ec01b849e2" withAuth=true
DEBU[0028] Request failed                                Status="Not Found" URL="http://<registry>/v2/e2e-automation/deploy/manifests/sha256-3e4b79029ce4089f26d78c8f3d98d936f1de7eb4051a251457b2a1ec01b849e2"
DEBU[0028] http req                                      method=HEAD url="http://<registry>/v2/e2e-automation/deploy/manifests/deploy" withAuth=true
DEBU[0029] http req                                      method=HEAD url="http://<registry>/v2/e2e-automation/deploy/manifests/deploy" withAuth=true
INFO[0029] Copy not needed, target already up to date    digest="sha256:3e4b79029ce4089f26d78c8f3d98d936f1de7eb4051a251457b2a1ec01b849e2" source="<registry>/e2e-automation/deploy:deploy" target="<registry>/e2e-automation/deploy:deploy"
<registry>/e2e-automation/deploy:deploy
```
### Expected Behavior
outputs with regctl v0.4.5, the real domain name has been hidden.

```
>>> regctl image mod --replace --annotation a=b <registry>/e2e-automation/deploy:deploy --verbosity debu
g
DEBU[0000] Loading docker config                         api= blobChunk=0 blobMax=0 helper=docker-credential-desktop hostname=registry-1.docker.io mirrors="[]" name=docker.io pathPrefix= repoAuth=false tls=enabled user=
DEBU[0000] Loading docker config                         api= blobChunk=0 blobMax=0 helper=docker-credential-desktop hostname=<registry> mirrors="[]" name=<registry> pathPrefix= repoAuth=false tls=enabled user=
DEBU[0000] Loading host config                           api= blobChunk=0 blobMax=0 helper= hostname=<registry> mirrors="[]" name=<registry> pathPrefix= repoAuth=false tls=disabled user=
WARN[0000] Changing TLS settings for registry            host=<registry> new=disabled orig=enabled
DEBU[0000] regclient initialized                         VCSRef=732c2a3a6312fc70dfcfcd821505ef0ad1ed230f VCSTag=v0.4.5
DEBU[0000] Modifying image                               ref="<registry>/e2e-automation/deploy:deploy"
DEBU[0000] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/manifests/deploy" withAuth=false
DEBU[0000] Auth request parsed                           challenge="[{bearer map[realm:http://<registry>/service/token scope:repository:e2e-automation/deploy:pull service:harbor-registry]}]"
DEBU[0000] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/manifests/deploy" withAuth=true
DEBU[0001] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/manifests/sha256:ab4c70e7d724f6de17e996fa2e8c0e01ae65802eae3b29b29591313c1f1063ce" withAuth=true
DEBU[0001] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/blobs/sha256:1b60b809c9a33d4410ffef20b04961eadd3cd4a669bd5cfe2a79ce8f76408031" withAuth=true
DEBU[0002] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/manifests/sha256:4a8227617ca11e623284852400c12cf2f30bf991597deac1abbc9f7d3927e22b" withAuth=true
DEBU[0003] http req                                      method=GET url="http://<registry>/v2/e2e-automation/deploy/blobs/sha256:ee9f4d7c388eafb922afbca879b1e3874e091a7b5a4e4be0546c93ce3a12b7cb" withAuth=true
DEBU[0003] http req                                      method=HEAD url="http://<registry>/v2/e2e-automation/deploy/manifests/deploy" withAuth=true
DEBU[0004] http req                                      method=HEAD url="http://<registry>/v2/e2e-automation/deploy/manifests/deploy" withAuth=true
INFO[0004] Copy not needed, target already up to date    digest="sha256:3e4b79029ce4089f26d78c8f3d98d936f1de7eb4051a251457b2a1ec01b849e2" source="<registry>/e2e-automation/deploy:deploy" target="<registry>/e2e-automation/deploy:deploy"
<registry>/e2e-automation/deploy:deploy
```

### Steps To Reproduce

run the same command with v0.4.7 and 0.4.5
```
regctl image mod --replace --annotation a=b <registry>/e2e-automation/deploy:deploy --verbosity debug
```

### Version

0.4.7
```text
>>> regctl version
VCSRef:     f43ada160d953317c51e7d9170afa2a316f4b9be
VCSCommit:  f43ada160d953317c51e7d9170afa2a316f4b9be
VCSState:   clean
VCSDate:    2023-02-26T02:43:41Z
Platform:   darwin/arm64
GoVer:      go1.20.1
GoCompiler: gc
```

0.4.5
```text
>>> regctl version
{
  "VCSRef": "732c2a3a6312fc70dfcfcd821505ef0ad1ed230f",
  "VCSTag": "v0.4.5"
}
```

### Environment

- Running as binary or container: binary
- Host platform: Mac M1 Pro ventura 13.2.1
- Registry description: Docker Hub

**Comment by sudo-bmitch** [comment]
Hi @airycanon. Please provide the output from:

```
curl -I -X GET http://<registry>/v2/e2e-automation/deploy/manifests/deploy
```

I'd like to see the `www-authenticate` header. Feel free to censor hostnames. Between the two releases, very little changed in the auth processing other than the handling of tokens that are about to expire, and I'm confused how that would cause this issue.

**Comment by sudo-bmitch** [comment]
Hi @airycanon, it looks like the referrers API is triggering some strange responses in the auth. From the logs I'm seeing that the registry is Harbor, do you happen to know the release/version of Harbor you are running? I'll do some testing here and see if there's a way to better handle their response.

**Comment by airycanon** [comment]
> Hi @airycanon, it looks like the referrers API is triggering some strange responses in the auth. From the logs I'm seeing that the registry is Harbor, do you happen to know the release/version of Harbor you are running? I'll do some testing here and see if there's a way to better handle their response.
Hi, thanks for your quick response, the version of Harbor is 2.2.3. 
Please let me know if you need any further information.

**Comment by sudo-bmitch** [comment]
Thanks. I've got this reproduced and a quick fix works. I'll have a PR pushed once I have some test cases and cleanup the logging.

**Comment by sudo-bmitch** [comment]
I believe this was fixed in #389. If you still have any issues with that, let me know and we can reopen this. If you are running regctl as a container, the `edge` tag has these fixes. For binaries, you can build from main or download from CI here: https://github.com/regclient/regclient/actions/runs/4399292427

Thanks for reporting this! I've also followed up with the Harbor devs since the HTTP response is unexpected.
